port/raspberrypi/rp2xxx/src/hal/i2c_slave.zig

```diff
@@ -0,0 +1,315 @@
+//! to be.
+//!
+//! Once the connection is open, and the appropriate ISR is set up,
+//! the driver will listen for I2C transactions.
+//!
+//! ## Master Write / Slave Read
+//!
+//! When we get data from the master we add it to the transfer buffer.  If the
+//! buffer becomes full and we have more data to add, we call the rx_callback
+//! with the `last` parameter set to `false`.
+//!
+//! When we have processed all the data from the master, and the master issues
+//! a STOP or RESTART condition, we call the rx_callback with the `last`
+//! parameter set to `true`.
+//!
+//! In either case, the `first` parameter will be set `true` for the initial
+//! callback for a transaction, and the `gen_call` parameter will be set to
+//! `true` for a general call and `false` if the transaction was addressed to us.
+//!
+//! ## Master Write / Slave Read
+//!
+//! If the master requests data from us we call the tx_callback with the `first`
+//! parameter set to `true`.  The user should add data to `data` and return the
+//! number of bytes added.
+//!
+//! If the master request more data, we call the tx_callback with the `first`
+//! parameter set to `false`.  The user can return 0 if no more data is available.
+//!
+
+const std = @import("std");
+const microzig = @import("microzig");
+const irq = @import("irq.zig");
+
+const mdf = microzig.drivers;
+const peripherals = microzig.chip.peripherals;
+const I2C0 = peripherals.I2C0;
+const I2C1 = peripherals.I2C1;
+const I2cRegs = microzig.chip.types.peripherals.I2C0;
+
+const fifo_length = 16;
+
+const gpio = @import("gpio.zig");
+
+pub const RXCallback = *const fn (data: []const u8, first: bool, last: bool, gen_call: bool, param: ?*anyopaque) void;
+pub const TXCallback = *const fn (data: []u8, first: bool, param: ?*anyopaque) usize;
+
+const Self = @This();
+
+var i2cs: [2]Self = .{ .{ .regs = I2C0 }, .{ .regs = I2C1 } };
+
+pub const i2c0 = &i2cs[0];
+pub const i2c1 = &i2cs[1];
+
+regs: *volatile I2cRegs = undefined,
+rxCallback: RXCallback = undefined,
+txCallback: TXCallback = undefined,
+param: ?*anyopaque = null,
+transfer_buffer: []u8 = undefined,
+transfer_index: usize = 0,
+transfer_length: usize = 0,
+data_received: bool = false,
+first_call: bool = true,
+gen_call: bool = false,
+tx_end: bool = false,
+
+pub fn isr0() callconv(.c) void {
+    i2cs[0].isr_common();
+}
+
+pub fn isr1() callconv(.c) void {
+    i2cs[1].isr_common();
+}
+
+/// Opens the I2C slave connection
+///
+/// ## Parameters
+///
+/// * `addr` - The I2C slave address to use
+/// * `transfer_buffer` - A buffer to use to transfer data
+/// * `rxCallback` - The callback to use for handling data received from i2c
+/// * `txCallback` - The callback to use for providing data sent to i2c
+/// * `param` - An optional parameter to pass to the callback
+pub fn open(self: *Self, addr: u7, transfer_buffer: []u8, rxCallback: RXCallback, txCallback: TXCallback, param: ?*anyopaque) void {
```

Comment:
Should use the `i2c.Address` from the master part of the hal

Reply:
Done!